lib/eco/export/jruby.py

```diff
@@ -0,0 +1,344 @@
+            self._export_as_text(path)
+            return
+        if run:
+            return self._run()
+        elif profile:
+            return self._profile()
+
+    def _language_box(self, name, node):
+        if name == '<Ruby>':
+            self._walk_rb(node)
+
+    def _walk_rb(self, node):
+        while True:
+            node = node.next_term
+            sym = node.symbol
+            if isinstance(node, EOS):
+                break
+            if isinstance(sym, MagicTerminal):
+                self._language_box(sym.name, node.symbol.ast.children[0])
+            elif isinstance(sym, IndentationTerminal):
+                self._output.append(sym)
+            elif sym.name == '\r':
+                self._output.append('\n')
+            else:
+                self._output.append(sym.name)
+
+    def _export_as_text(self, path):
+        node = self.tm.lines[0].node # first node
+        self._walk_rb(node)
+        output = ''.join(self._output)
+        with open(path, 'w') as fp:
+            fp.write(''.join(output))
+
+    def _run(self):
+        f = tempfile.mkstemp(suffix='.rb')
+        settings = QSettings('softdev', 'Eco')
+        jruby_bin =str (settings.value('env_jruby').toString())
+        self._export_as_text(f[1])
+        # Run this command:
+        #     $ jruby -X+T FILE.rb
+        return subprocess.Popen([jruby_bin,
+                                 '-X+T',
+                                 f[1]],
+                                stdout=subprocess.PIPE,
+                                stderr=subprocess.STDOUT,
+                                bufsize=0)
+
+    def _annotate_text(self, lineno, text, annotation, klass):
+        """Annotate a node on a given line with a given symbol name.
+        FIXME: Should handle case where line (and all subsequent lines) have
+        already been deleted by the user.
+        FIXME: Needs to handle the case where same text appears in more than
+        one node on the same line.
+        """
+        try:
+            temp_cursor = self.tm.cursor.copy()
+            temp_cursor.line = lineno - 1
+            temp_cursor.move_to_x(0, self.tm.lines)
+            node = temp_cursor.find_next_visible(temp_cursor.node)
+            while node.lookup == '<ws>' or node.symbol.name != text:
+                node = node.next_term
+                if isinstance(node, EOS):
+                    raise ValueError('EOS')
+            node.remove_annotations_by_class(klass)
+            node.add_annotation(klass(annotation))
+        except ValueError:
+            pass
+
+    def _profile(self):
+        src_file_fd, src_file_name = tempfile.mkstemp(suffix='.rb')
+        self.tm.export_as_text(src_file_name).split('\n')
+
+        info_file_name = os.path.join('/',
+                                      'tmp',
+                                      next(tempfile._get_candidate_names()) + '.txt')
+        print 'Placing callgraph trace in', info_file_name
```

Comment:
I've been using the `logging` module. Most of the time I'm printing messages through it's `debug` method, e.g. `logging.debug("some string: %s", s)`. Debug messages can then be activated running Eco with `eco.py --log=debug`.